spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/integration/outbound/PubSubMessageHandler.java

```diff
@@ -82,30 +86,35 @@ protected void handleMessageInternal(Message<?> message) throws Exception {
 				? message.getHeaders().get(GcpPubSubHeaders.TOPIC, String.class)
 				: this.topicExpression.getValue(this.evaluationContext, message, String.class);
 
-		if (payload instanceof PubsubMessage) {
-			this.pubSubTemplate.publish(topic, (PubsubMessage) payload);
-			return;
-		}
+		ListenableFuture<String> pubsubFuture;
 
-		ByteString pubsubPayload;
-
-		if (payload instanceof byte[]) {
-			pubsubPayload = ByteString.copyFrom((byte[]) payload);
-		}
-		else if (payload instanceof ByteString) {
-			pubsubPayload = (ByteString) payload;
+		if (payload instanceof PubsubMessage) {
+			pubsubFuture = this.pubSubTemplate.publish(topic, (PubsubMessage) payload);
 		}
 		else {
-			pubsubPayload =	ByteString.copyFrom(
-					(String) this.messageConverter.fromMessage(message, String.class),
-					Charset.defaultCharset());
-		}
+			ByteString pubsubPayload;
 
-		Map<String, String> headers = new HashMap<>();
-		this.headerMapper.fromHeaders(message.getHeaders(), headers);
+			if (this.messageConverter != null) {
+				pubsubPayload = ByteString.copyFrom(
+						(byte[]) this.messageConverter.fromMessage(message, byte[].class));
+			}
+			else if (payload instanceof byte[]) {
+				pubsubPayload = ByteString.copyFrom((byte[]) payload);
+			}
+			else if (payload instanceof ByteString) {
+				pubsubPayload = (ByteString) payload;
+			}
+			else {
+				throw new PubSubException("Unable to convert payload of type "
```

Comment:
Sounds like a `MessageConversionException`? Seems for me it's a bit early to fail as a Pub/Sub.

But that's IMO I guess.

Reply:
Sure, that works.
I'll update the refdoc tomorrow.